File: airbyte-db/db-lib/src/main/java/io/airbyte/db/jdbc/DateTimeConverter.java
/*
 * Copyright (c) 2022 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.db.jdbc;

import static io.airbyte.db.DataTypeUtils.DATE_FORMATTER;
import static io.airbyte.db.DataTypeUtils.TIMESTAMPTZ_FORMATTER;
import static io.airbyte.db.DataTypeUtils.TIMESTAMP_FORMATTER;
import static io.airbyte.db.DataTypeUtils.TIMETZ_FORMATTER;
import static io.airbyte.db.DataTypeUtils.TIME_FORMATTER;
import static io.airbyte.db.jdbc.AbstractJdbcCompatibleSourceOperations.resolveEra;
import static java.time.ZoneOffset.UTC;

import com.fasterxml.jackson.databind.node.ObjectNode;
import java.sql.Date;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Time;
import java.sql.Timestamp;
import java.time.Duration;
import java.time.Instant;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.LocalTime;
import java.time.OffsetDateTime;
import java.time.OffsetTime;
import java.time.ZonedDateTime;
import java.time.format.DateTimeFormatter;
import java.util.concurrent.TimeUnit;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class DateTimeConverter {

  private static final Logger LOGGER = LoggerFactory.getLogger(DateTimeConverter.class);
  public static final DateTimeFormatter TIME_WITH_TIMEZONE_FORMATTER = DateTimeFormatter.ofPattern(
      "HH:mm:ss[.][SSSSSSSSS][SSSSSSS][SSSSSS][SSSSS][SSSS][SSS][SS][S][''][XXX][XX][X]");
  private static boolean loggedUnknownTimeWithTimeZoneClass = false;
  private static boolean loggedUnknownTimeClass = false;
  private static boolean loggedUnknownTimestampWithTimeZoneClass = false;
  private static boolean loggedUnknownTimestampClass = false;
  private static boolean loggedUnknownDateClass = false;

  public static String convertToTimeWithTimezone(final Object time) {
    if (time instanceof final java.time.OffsetTime timetz) {
      return timetz.format(TIMETZ_FORMATTER);
    } else {
      if (!loggedUnknownTimeWithTimeZoneClass) {
        LOGGER.info("Unknown class for Time with timezone data type" + time.getClass());
        loggedUnknownTimeWithTimeZoneClass = true;
      }
      final OffsetTime timetz = OffsetTime.parse(time.toString(), TIME_WITH_TIMEZONE_FORMATTER);
      return timetz.format(TIMETZ_FORMATTER);
    }
  }

  public static String convertToTimestampWithTimezone(final Object timestamp) {
    if (timestamp instanceof final Timestamp t) {
      // In snapshot mode, debezium produces a java.sql.Timestamp object for the TIMESTAMPTZ type.
      // Conceptually, a timestamp with timezone is an Instant. But t.toInstant() actually mangles the
      // value for ancient dates, because leap years weren't applied consistently in ye olden days.
      // Additionally, toInstant() (and toLocalDateTime()) actually lose the era indicator, so we can't
      // rely on their getEra() methods.
      // So we have special handling for this case, which sidesteps the toInstant conversion.
      final ZonedDateTime timestamptz = t.toLocalDateTime().atZone(UTC);
      final String value = timestamptz.format(TIMESTAMPTZ_FORMATTER);
      return resolveEra(t, value);
    } else if (timestamp instanceof final OffsetDateTime t) {
      return resolveEra(t.toLocalDate(), t.format(TIMESTAMPTZ_FORMATTER));
    } else if (timestamp instanceof final ZonedDateTime timestamptz) {
      return resolveEra(timestamptz.toLocalDate(), timestamptz.format(TIMESTAMPTZ_FORMATTER));
    } else if (timestamp instanceof final Instant instant) {
      final OffsetDateTime offsetDateTime = OffsetDateTime.ofInstant(instant, UTC);
      final ZonedDateTime timestamptz = ZonedDateTime.from(offsetDateTime);
      final LocalDate localDate = timestamptz.toLocalDate();
      final String value = timestamptz.format(TIMESTAMPTZ_FORMATTER);
      return resolveEra(localDate, value);
    } else {
      if (!loggedUnknownTimestampWithTimeZoneClass) {
        LOGGER.info("Unknown class for Timestamp with time zone data type" + timestamp.getClass());
        loggedUnknownTimestampWithTimeZoneClass = true;
      }
      final Instant instant = Instant.parse(timestamp.toString());
      final OffsetDateTime offsetDateTime = OffsetDateTime.ofInstant(instant, UTC);
      final ZonedDateTime timestamptz = ZonedDateTime.from(offsetDateTime);
      final LocalDate localDate = timestamptz.toLocalDate();
      final String value = timestamptz.format(TIMESTAMPTZ_FORMATTER);
      return resolveEra(localDate, value);
    }
  }

  /**
   * See {@link #convertToTimestampWithTimezone(Object)} for explanation of the weird things happening
   * here.
   */
  public static String convertToTimestamp(final Object timestamp) {
    if (timestamp instanceof final Timestamp t) {
      // Snapshot mode
      final LocalDateTime localDateTime = t.toLocalDateTime();
      final String value = localDateTime.format(TIMESTAMP_FORMATTER);
      return resolveEra(t, value);
    } else if (timestamp instanceof final Instant i) {
      // Incremental mode
      return resolveEra(i.atZone(UTC).toLocalDate(), i.atOffset(UTC).toLocalDateTime().format(TIMESTAMP_FORMATTER));
    } else if (timestamp instanceof final LocalDateTime localDateTime) {
      final LocalDate date = localDateTime.toLocalDate();
      final String value = localDateTime.format(TIMESTAMP_FORMATTER);
      return resolveEra(date, value);
    } else {
      if (!loggedUnknownTimestampClass) {
        LOGGER.info("Unknown class for Timestamp data type" + timestamp.getClass());
        loggedUnknownTimestampClass = true;
      }
      final LocalDateTime localDateTime = LocalDateTime.parse(timestamp.toString());
      final LocalDate date = localDateTime.toLocalDate();
      final String value = localDateTime.format(TIMESTAMP_FORMATTER);
      return resolveEra(date, value);
    }
  }

  /**
   * See {@link #convertToTimestampWithTimezone(Object)} for explanation of the weird things happening
   * here.
   */
  public static String convertToDate(final Object date) {
    if (date instanceof final Date d) {
      // Snapshot mode
      final LocalDate localDate = ((Date) date).toLocalDate();
      return resolveEra(d, localDate.format(DATE_FORMATTER));
    } else if (date instanceof LocalDate d) {
      // Incremental mode
      return resolveEra(d, d.format(DATE_FORMATTER));
    } else {
      if (!loggedUnknownDateClass) {
        LOGGER.info("Unknown class for Date data type" + date.getClass());
        loggedUnknownDateClass = true;
      }
      final LocalDate localDate = LocalDate.parse(date.toString());
      return resolveEra(localDate, localDate.format(DATE_FORMATTER));
    }
  }

  public static String convertToTime(final Object time) {
    if (time instanceof final Time sqlTime) {
      return sqlTime.toLocalTime().format(TIME_FORMATTER);
    } else if (time instanceof final LocalTime localTime) {
      return localTime.format(TIME_FORMATTER);
    } else if (time instanceof java.time.Duration) {
      long value = ((Duration) time).toNanos();
      if (value >= 0 && value < TimeUnit.DAYS.toNanos(1)) {
        return LocalTime.ofNanoOfDay(value).format(TIME_FORMATTER);
      } else {
        final long updatedValue = Math.min(Math.abs(value), LocalTime.MAX.toNanoOfDay());
        LOGGER.debug("Time values must use number of nanoseconds greater than 0 and less than 86400000000000 but its {}, converting to {} ", value,
            updatedValue);
        return LocalTime.ofNanoOfDay(updatedValue).format(TIME_FORMATTER);
      }
    } else {
      if (!loggedUnknownTimeClass) {
        LOGGER.info("Unknown class for Time data type" + time.getClass());
        loggedUnknownTimeClass = true;
      }

      final String valueAsString = time.toString();
      if (valueAsString.startsWith("24")) {
        LOGGER.debug("Time value {} is above range, converting to 23:59:59", valueAsString);
        return LocalTime.MAX.format(TIME_FORMATTER);
      }
      return LocalTime.parse(valueAsString).format(TIME_FORMATTER);
    }
  }

  public static void putJavaSQLDate(ObjectNode node, String columnName, ResultSet resultSet, int index) throws SQLException {
    final Date date = resultSet.getDate(index);
    node.put(columnName, convertToDate(date));
  }

  public static void putJavaSQLTime(ObjectNode node, String columnName, ResultSet resultSet, int index) throws SQLException {
    // resultSet.getTime() will lose nanoseconds precision
    final LocalTime localTime = resultSet.getTimestamp(index).toLocalDateTime().toLocalTime();
    node.put(columnName, convertToTime(localTime));
  }

}

File: airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/db/jdbc/DateTimeConverter.kt
/*
 * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
 */
package io.airbyte.cdk.db.jdbc

import com.fasterxml.jackson.databind.node.ObjectNode
import io.airbyte.cdk.db.DataTypeUtils
import java.sql.*
import java.time.*
import java.time.format.DateTimeFormatter
import java.util.concurrent.*
import kotlin.math.abs
import kotlin.math.min
import org.slf4j.Logger
import org.slf4j.LoggerFactory

object DateTimeConverter {
    private val LOGGER: Logger = LoggerFactory.getLogger(DateTimeConverter::class.java)
    val TIME_WITH_TIMEZONE_FORMATTER: DateTimeFormatter =
        DateTimeFormatter.ofPattern(
            "HH:mm:ss[.][SSSSSSSSS][SSSSSSS][SSSSSS][SSSSS][SSSS][SSS][SS][S][''][XXX][XX][X]"
        )
    private var loggedUnknownTimeWithTimeZoneClass = false
    private var loggedUnknownTimeClass = false
    private var loggedUnknownTimestampWithTimeZoneClass = false
    private var loggedUnknownTimestampClass = false
    private var loggedUnknownDateClass = false

    @JvmStatic
    fun convertToTimeWithTimezone(time: Any): String {
        if (time is OffsetTime) {
            return if (hasZeroSecondsAndNanos(time.toLocalTime()))
                time.format(DataTypeUtils.TIMETZ_FORMATTER)
            else time.toString()
        } else {
            if (!loggedUnknownTimeWithTimeZoneClass) {
                LOGGER.info("Unknown class for Time with timezone data type" + time.javaClass)
                loggedUnknownTimeWithTimeZoneClass = true
            }
            val timetz = OffsetTime.parse(time.toString(), TIME_WITH_TIMEZONE_FORMATTER)
            return if (hasZeroSecondsAndNanos(timetz.toLocalTime()))
                timetz.format(DataTypeUtils.TIMETZ_FORMATTER)
            else timetz.toString()
        }
    }

    @JvmStatic
    fun convertToTimestampWithTimezone(timestamp: Any): String {
        if (timestamp is Timestamp) {
            // In snapshot mode, debezium produces a java.sql.Timestamp object for the TIMESTAMPTZ
            // type.
            // Conceptually, a timestamp with timezone is an Instant. But t.toInstant() actually
            // mangles the
            // value for ancient dates, because leap years weren't applied consistently in ye olden
            // days.
            // Additionally, toInstant() (and toLocalDateTime()) actually lose the era indicator, so
            // we can't
            // rely on their getEra() methods.
            // So we have special handling for this case, which sidesteps the toInstant conversion.
            val timestamptz: ZonedDateTime = timestamp.toLocalDateTime().atZone(ZoneOffset.UTC)
            val value = timestamptz.format(DataTypeUtils.TIMESTAMPTZ_FORMATTER)
            return AbstractJdbcCompatibleSourceOperations.Companion.resolveEra(timestamp, value)
        } else if (timestamp is OffsetDateTime) {
            return AbstractJdbcCompatibleSourceOperations.Companion.resolveEra(
                timestamp.toLocalDate(),
                timestamp.format(DataTypeUtils.TIMESTAMPTZ_FORMATTER)
            )
        } else if (timestamp is ZonedDateTime) {
            return AbstractJdbcCompatibleSourceOperations.Companion.resolveEra(
                timestamp.toLocalDate(),
                timestamp.format(DataTypeUtils.TIMESTAMPTZ_FORMATTER)
            )
        } else if (timestamp is Instant) {
            val offsetDateTime = OffsetDateTime.ofInstant(timestamp, ZoneOffset.UTC)
            val timestamptz = ZonedDateTime.from(offsetDateTime)
            val localDate = timestamptz.toLocalDate()
            val value = timestamptz.format(DataTypeUtils.TIMESTAMPTZ_FORMATTER)
            return AbstractJdbcCompatibleSourceOperations.Companion.resolveEra(localDate, value)
        } else {
            if (!loggedUnknownTimestampWithTimeZoneClass) {
                LOGGER.info(
                    "Unknown class for Timestamp with time zone data type" + timestamp.javaClass
                )
                loggedUnknownTimestampWithTimeZoneClass = true
            }
            val instant = Instant.parse(timestamp.toString())
            val offsetDateTime = OffsetDateTime.ofInstant(instant, ZoneOffset.UTC)
            val timestamptz = ZonedDateTime.from(offsetDateTime)
            val localDate = timestamptz.toLocalDate()
            val value = timestamptz.format(DataTypeUtils.TIMESTAMPTZ_FORMATTER)
            return AbstractJdbcCompatibleSourceOperations.Companion.resolveEra(localDate, value)
        }
    }

    /** See [.convertToTimestampWithTimezone] for explanation of the weird things happening here. */
    @JvmStatic
    fun convertToTimestamp(timestamp: Any): String {
        if (timestamp is Timestamp) {
            // Snapshot mode
            val localDateTime: LocalDateTime = timestamp.toLocalDateTime()
            return AbstractJdbcCompatibleSourceOperations.Companion.resolveEra(
                timestamp,
                if (hasZeroSecondsAndNanos(localDateTime.toLocalTime()))
                    localDateTime.format(DataTypeUtils.TIMESTAMP_FORMATTER)
                else localDateTime.toString()
            )
        } else if (timestamp is Instant) {
            // Incremental mode
            return AbstractJdbcCompatibleSourceOperations.Companion.resolveEra(
                timestamp.atZone(ZoneOffset.UTC).toLocalDate(),
                timestamp
                    .atOffset(ZoneOffset.UTC)
                    .toLocalDateTime()
                    .format(DataTypeUtils.TIMESTAMP_FORMATTER)
            )
        } else if (timestamp is LocalDateTime) {
            val date: LocalDate = timestamp.toLocalDate()
            return AbstractJdbcCompatibleSourceOperations.Companion.resolveEra(
                date,
                if (hasZeroSecondsAndNanos(timestamp.toLocalTime()))
                    timestamp.format(DataTypeUtils.TIMESTAMP_FORMATTER)
                else timestamp.toString()
            )
        } else {
            if (!loggedUnknownTimestampClass) {
                LOGGER.info("Unknown class for Timestamp data type" + timestamp.javaClass)
                loggedUnknownTimestampClass = true
            }
            val localDateTime = LocalDateTime.parse(timestamp.toString())
            val date = localDateTime.toLocalDate()
            return AbstractJdbcCompatibleSourceOperations.Companion.resolveEra(
                date,
                if (hasZeroSecondsAndNanos(localDateTime.toLocalTime()))
                    localDateTime.format(DataTypeUtils.TIMESTAMP_FORMATTER)
                else localDateTime.toString()
            )
        }
    }

    /** See [.convertToTimestampWithTimezone] for explanation of the weird things happening here. */
    @JvmStatic
    fun convertToDate(date: Any): String {
        if (date is Date) {
            // Snapshot mode
            val localDate = date.toLocalDate()
            return AbstractJdbcCompatibleSourceOperations.Companion.resolveEra(
                date,
                localDate.format(DataTypeUtils.DATE_FORMATTER)
            )
        } else if (date is LocalDate) {
            // Incremental mode
            return AbstractJdbcCompatibleSourceOperations.Companion.resolveEra(
                date,
                date.format(DataTypeUtils.DATE_FORMATTER)
            )
        } else if (date is Integer) {
            // Incremental mode
            return LocalDate.ofEpochDay(date.toLong()).format(DataTypeUtils.DATE_FORMATTER)
        } else {
            if (!loggedUnknownDateClass) {
                LOGGER.info("Unknown class for Date data type" + date.javaClass)
                loggedUnknownDateClass = true
            }
            val localDate = LocalDate.parse(date.toString())
            return AbstractJdbcCompatibleSourceOperations.Companion.resolveEra(
                localDate,
                localDate.format(DataTypeUtils.DATE_FORMATTER)
            )
        }
    }

    @JvmStatic
    fun convertToTime(time: Any): String {
        if (time is Time) {
            return formatTime(time.toLocalTime())
        } else if (time is LocalTime) {
            return formatTime(time)
        } else if (time is Duration) {
            val value = time.toNanos()
            if (value >= 0 && value < TimeUnit.DAYS.toNanos(1)) {
                return formatTime(LocalTime.ofNanoOfDay(value))
            } else {
                val updatedValue =
                    min(abs(value.toDouble()), LocalTime.MAX.toNanoOfDay().toDouble()).toLong()
                LOGGER.debug(
                    "Time values must use number of nanoseconds greater than 0 and less than 86400000000000 but its {}, converting to {} ",
                    value,
                    updatedValue
                )
                return formatTime(LocalTime.ofNanoOfDay(updatedValue))
            }
        } else {
            if (!loggedUnknownTimeClass) {
                LOGGER.info("Unknown class for Time data type" + time.javaClass)
                loggedUnknownTimeClass = true
            }

            val valueAsString = time.toString()
            if (valueAsString.startsWith("24")) {
                LOGGER.debug("Time value {} is above range, converting to 23:59:59", valueAsString)
                return LocalTime.MAX.toString()
            }
            return formatTime(LocalTime.parse(valueAsString))
        }
    }

    @JvmStatic
    private fun formatTime(localTime: LocalTime): String {
        return if (hasZeroSecondsAndNanos(localTime)) localTime.format(DataTypeUtils.TIME_FORMATTER)
        else localTime.toString()
    }

    @JvmStatic
    fun hasZeroSecondsAndNanos(localTime: LocalTime): Boolean {
        return (localTime.second == 0 && localTime.nano == 0)
    }

    @Throws(SQLException::class)
    @JvmStatic
    fun putJavaSQLDate(node: ObjectNode, columnName: String?, resultSet: ResultSet, index: Int) {
        val date = resultSet.getDate(index)
        node.put(columnName, convertToDate(date))
    }

    @Throws(SQLException::class)
    @JvmStatic
    fun putJavaSQLTime(node: ObjectNode, columnName: String?, resultSet: ResultSet, index: Int) {
        // resultSet.getTime() will lose nanoseconds precision
        val localTime = resultSet.getTimestamp(index).toLocalDateTime().toLocalTime()
        node.put(columnName, convertToTime(localTime))
    }
}
